### PR TITLE
Fix isolated process using split APKs

### PR DESCRIPTION
**Pull request by JingMatrix**
Solve issue #190 : for isolated process, it is impossible to extract original APK from assets.

Currently, we still need to work on the `android:zygotePreloadName` problem, which causes that `ActivityManager` fails for isolated processes having the  `android:useAppZygote` attribute .

See docs on [ZygotePreload](https://developer.android.com/reference/android/app/ZygotePreload).

### CODE REVIEW

**Comment by JingMatrix** [comment]
In the second commit, I found that `ActivityThread.currentActivityThread()` could return null and then crashed the app.

This happens for `AppZygoteInit.java`, see below log for details
```
05-29 21:05:46.028 19138 19138 I LSPatch-MetaLoader: Bootstrap loader from embedment
05-29 21:05:46.038 19138 19138 D LSPatch-MetaLoader: ActivityThread.currentActivityThread is null, dumping stack trace
05-29 21:05:46.038 19138 19138 D LSPatch-MetaLoader: [dalvik.system.VMStack.getThreadStackTrace(Native Method)
05-29 21:05:46.038 19138 19138 D LSPatch-MetaLoader:  java.lang.Thread.getStackTrace(Thread.java:1841)
05-29 21:05:46.038 19138 19138 D LSPatch-MetaLoader:  org.lsposed.lspatch.metaloader.LSPAppComponentFactoryStub.<clinit>(LSPAppComponentFactoryStub.java:99)
05-29 21:05:46.038 19138 19138 D LSPatch-MetaLoader:  java.lang.Class.newInstance(Native Method)
05-29 21:05:46.038 19138 19138 D LSPatch-MetaLoader:  android.app.LoadedApk.createAppFactory(LoadedApk.java:273)
05-29 21:05:46.038 19138 19138 D LSPatch-MetaLoader:  android.app.LoadedApk.createOrUpdateClassLoaderLocked(LoadedApk.java:1039)
05-29 21:05:46.038 19138 19138 D LSPatch-MetaLoader:  android.app.LoadedApk.getClassLoader(LoadedApk.java:1126)
05-29 21:05:46.038 19138 19138 D LSPatch-MetaLoader:  com.android.internal.os.AppZygoteInit$AppZygoteConnection.handlePreloadApp(AppZygoteInit.java:77)
05-29 21:05:46.038 19138 19138 D LSPatch-MetaLoader:  com.android.internal.os.ZygoteConnection.processCommand(ZygoteConnection.java:182)
05-29 21:05:46.038 19138 19138 D LSPatch-MetaLoader:  com.android.internal.os.ZygoteServer.runSelectLoop(ZygoteServer.java:552)
05-29 21:05:46.038 19138 19138 D LSPatch-MetaLoader:  com.android.internal.os.ChildZygoteInit.runZygoteServer(ChildZygoteInit.java:125)
05-29 21:05:46.038 19138 19138 D LSPatch-MetaLoader:  com.android.internal.os.AppZygoteInit.main(AppZygoteInit.java:119)
05-29 21:05:46.038 19138 19138 D LSPatch-MetaLoader:  java.lang.reflect.Method.invoke(Native Method)
05-29 21:05:46.038 19138 19138 D LSPatch-MetaLoader:  com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:548)
05-29 21:05:46.038 19138 19138 D LSPatch-MetaLoader:  com.android.internal.os.ZygoteInit.main(ZygoteInit.java:936)]
```

**Comment by JingMatrix** [comment]
The main difficulty is to change `appInfo` for `AppZygoteInit`, which is defined in [ZygoteConnection.java](https://cs.android.com/android/platform/superproject/+/refs/heads/master:frameworks/base/core/java/com/android/internal/os/ZygoteConnection.java;l=178-182;drc=7346c436e5a11ce08f6a80dcfeb8ef941ca30176;bpv=0;bpt=1j), and then used in [AppZygoteInit.java](https://cs.android.com/android/platform/superproject/+/refs/heads/master:frameworks/base/core/java/com/android/internal/os/AppZygoteInit.java;l=76-77;bpv=0;bpt=1).

This time we cant not reflect `ActivityThread.java` to change it.
 We should make a [ZygotePreload](https://developer.android.com/reference/android/app/ZygotePreload) stub, and change `appInfo` in its method `doPreload`, which is called in [AppZygoteInit.java](https://cs.android.com/android/platform/superproject/+/refs/heads/master:frameworks/base/core/java/com/android/internal/os/AppZygoteInit.java;l=95;bpv=0;bpt=1).

**Comment by JingMatrix** [comment]
Using the method in my previous comment, the problem is fully solved.
Chrome is fully supported without any problems.

Please test and merge. (If possible, I would be happy to be a maintainer of `LSPatch`.)

**Comment by yujincheng08** [comment]
What's the indent going on?

File: meta-loader/build.gradle.kts [hunk 1]
```diff
@@ -44,4 +44,5 @@ dependencies {
     compileOnly(projects.hiddenapi.stubs)
     implementation(projects.share.java)
     implementation("org.lsposed.hiddenapibypass:hiddenapibypass:4.3")
+    implementation("com.google.code.gson:gson:2.10")
```
Comment (yujincheng08) [inline meta-loader/build.gradle.kts:47]: We don't need gson. Use JsonObject instead.

File: gradle.properties [hunk 2]
```diff
@@ -6,3 +6,5 @@ android.enableR8.fullMode=true
 android.useAndroidX=true
 
 agpVersion=7.4.0
+
+org.gradle.jvmargs=-Xmx2048m
```
Comment (yujincheng08) [inline gradle.properties:10]: dont write it here.

File: patch-loader/src/main/java/org/lsposed/lspatch/loader/LSPApplication.java [hunk 3]
```diff
@@ -65,16 +65,15 @@ public static boolean isIsolated() {
...
-            XLog.d(TAG, "Skip isolated process");
-            return;
-        }
         activityThread = ActivityThread.currentActivityThread();
         var context = createLoadedApkWithContext();
-        if (context == null) {
-            XLog.e(TAG, "Error when creating context");
+
+        if (isIsolated()) {
+			XLog.d(TAG, "Skip isolated process");
```
Comment (yujincheng08) [inline patch-loader/src/main/java/org/lsposed/lspatch/loader/LSPApplication.java:72]: still tab

File: meta-loader/src/main/java/org/lsposed/lspatch/metaloader/LSPZygotePreloadStub.java [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+            Class<?> originalPreload = cl.loadClass(config.getString("zygotePreloadName"));
+            Method originalDoPreload = originalPreload.getDeclaredMethod("doPreload", ApplicationInfo.class);
+            originalDoPreload.invoke(originalPreload.getDeclaredConstructor().newInstance(), appInfo);
+        } catch (Throwable e) {
+            Log.d(TAG, "Fail to call original ZygotePreload", e);
+        }
+    }
+}
+
+/* vim: set tabstop=4 expandtab: */
```
Comment (yujincheng08) [inline meta-loader/src/main/java/org/lsposed/lspatch/metaloader/LSPZygotePreloadStub.java:51]: remove this

**Comment by JingMatrix** [comment]
Thanks for reviewing.
Tab indent problem is fixed (I forgot to adjust my vim settings) and gson is removed.

By the way, I think this method using split APKs remove the need to `switchAllClassLoader` because both `base` and `split_original` apks are in the `DexPathList`.
Should we better simplify the logic in `LSPApplication.java`?

PS: Modifications on the manager app are not finished yet.

File: patch-loader/src/main/java/org/lsposed/lspatch/loader/LSPApplication.java [hunk 5]
```diff
@@ -119,25 +118,12 @@ private static Context createLoadedApkWithContext() {
...
-            }
+            String originalApkPath = Paths.get(appInfo.sourceDir).getParent().toString() + "/split_original.apk";
 
-            appInfo.sourceDir = cacheApkPath.toString();
-            appInfo.publicSourceDir = cacheApkPath.toString();
+            appInfo.sourceDir = originalApkPath;
+            appInfo.publicSourceDir = originalApkPath;
             appInfo.appComponentFactory = config.appComponentFactory;
 
-            if (!Files.exists(cacheApkPath)) {
```
Comment (yujincheng08) [inline patch-loader/src/main/java/org/lsposed/lspatch/loader/LSPApplication.java:132]: ?

File: patch/src/main/java/org/lsposed/patch/LSPatch.java [hunk 6]
```diff
@@ -281,26 +301,14 @@ public void patch(File srcApkFile, File outputFile) throws PatchError, IOExcepti
                 }
 
                 logger.i("Embedding modules...");
-                embedModules(dstZFile);
+                embedModules(baseZFile);
             }
 
-            // create zip link
-            logger.d("Creating nested apk link...");
```
Comment (yujincheng08) [inline patch/src/main/java/org/lsposed/patch/LSPatch.java:288]: ?

**Comment by yujincheng08** [comment]
You removed ziplink. why?

File: gradle.properties (hunk 2 above)
Comment (JingMatrix) [reply gradle.properties]: I encounter the same problem as this [StackOverflow question](https://stackoverflow.com/a/70756876/7870953).
Subproject setting is not possible, so I set it only for my local machine.

**Comment by yujincheng08** [comment]
hmm, I somehow know your idea. you also produce two apks after patching. but hv u think about embedded mode? ppl may distribute their patched apk to others. if lspatch always produces two apk, it's not easy for other to install. so this is not a feasible solution.

**Comment by yujincheng08** [comment]
有 isolated 的 app 不一定都是 split apk。这样用户没法在内嵌模块模式下分发打包好的 apk。因为安装 split apk 对很多人来说非常困难。所以 lspatch 不可能抛弃这种打包方式。

**Comment by yujincheng08** [comment]
其实关键问题就是 isolated process 没有可写目录用来临时存放原始 apk。但是不一定需要存原始 apk，我们可以把原始 apk 的 dex 读到内存，然后使用 InMemoryDexClassLoader，这样把原来的 classloader 变出来，但是这样签名是过不了的。

另一个思路是把它放 apk!/lib 下面，然后让系统把原来的 apk 解压到 lib，我不知道 isolated process 不知道能不能访问这个目录的文件，我觉得大概率是可以的。

由于有 isolated process 的 app 不是很多，我建议可以先尝试第二种。

**Comment by JingMatrix** [comment]
I will try to make a new pull-request to first fix the zygotePreloadName problem.

I understand your concern for multiple APKs. Thanks for sharing your ideas. I will try them.

**Comment by JingMatrix** [comment]
Nope, `apk!/lib` is not accessible for isolated processes, not even for normal processes:

```
05-30 12:42:15.295  9011  9011 I LSPatch-MetaLoader: Bootstrap loader from embedment
05-30 12:42:15.302  9011  9011 D LSPatch-MetaLoader: LSPAppComponentFactoryStub for process: org.bromite.bromite
05-30 12:42:15.326  9011  9011 I LSPatch : Use manager: false
05-30 12:42:15.326  9011  9011 I LSPatch : Signature bypass level: 0
05-30 12:42:15.326  9011  9011 D LSPatch : No access to /data/app/~~WrmcWdTYSMXTMfijfU9LSg==/org.bromite.bromite-5CW3ibYQahr7vdU67EQkTA==/base.apk!/lib/arm64-v8a
05-30 12:42:15.328  9011  9011 I LSPatch : hooked app initialized: android.app.LoadedApk@e398eec
05-30 12:42:15.330  9011  9011 W LSPatch : Original AppComponentFactory not found: org.chromium.chrome.browser.base.SplitCompatAppComponentFactory
05-30 12:42:15.330  9011  9011 D LSPatch : Initialize service client
05-30 12:42:15.334  9011  9011 D LSPatch : Processing /data/misc/profiles/cur/0/org.bromite.bromite/primary.prof
05-30 12:42:15.334  9011  9011 D LSPatch : Skip profile /data/misc/profiles/cur/0/org.bromite.bromite/primary.prof
05-30 12:42:15.336  9011  9011 I LSPatch : Bootstrap Xposed
05-30 12:42:15.340  9011  9011 I LSPatch : Load modules
05-30 12:42:15.376  9011  9011 I LSPatch : Modules initialized
05-30 12:42:15.376  9011  9011 I LSPatch : LSPatch bootstrap completed
05-30 12:42:15.428  9011  9011 I ChromeXt: Package: org.bromite.bromite, v108.0.5359.156
05-30 12:42:15.433  9011  9011 I ChromeXt: Inited hook: m
05-30 12:42:15.435  9011  9011 I ChromeXt: Inited hook: e
05-30 12:42:15.439  9011  9011 I ChromeXt: Inited hook: j
05-30 12:42:15.440  9011  9011 I ChromeXt: Inited hook: f
05-30 12:42:15.587  9055  9055 I LSPatch-MetaLoader: Bootstrap loader from embedment
05-30 12:42:15.598  9055  9055 D LSPatch-MetaLoader: Skip LSPAppComponentFactoryStub since AppZygoteInit is running
05-30 12:42:15.600  9055  9055 D LSPatch-MetaLoader: No access to /data/app/~~WrmcWdTYSMXTMfijfU9LSg==/org.bromite.bromite-5CW3ibYQahr7vdU67EQkTA==/base.apk!/lib/arm64-v8a
05-30 12:42:15.600  9055  9055 D LSPatch-MetaLoader: APK not found
05-30 12:42:15.600  9055  9055 D LSPatch-MetaLoader: Fail to call original ZygotePreload
05-30 12:42:15.600  9055  9055 D LSPatch-MetaLoader: java.lang.ClassNotFoundException: Didn't find class "org.chromium.content_public.app.ZygotePreload" on path: DexPathList[[zip file "/data/app/~~WrmcWdTYSMXTMfijfU9LSg==/org.bromite.bromite-5CW3ibYQahr7vdU67EQkTA==/base.apk"],nativeLibraryDirectories=[/data/app/~~WrmcWdTYSMXTMfijfU9LSg==/org.bromite.bromite-5CW3ibYQahr7vdU67EQkTA==/lib/arm64, /data/app/~~WrmcWdTYSMXTMfijfU9LSg==/org.bromite.bromite-5CW3ibYQahr7vdU67EQkTA==/base.apk!/lib/arm64-v8a, /system/lib64, /system_ext/lib64]]
```


I think we can add an option to the comand line `lspatch.jar` to merge the split APKs, such projects seem already existed.

Or, we leave the users to choose whether they would use split APKs.

**Comment by yujincheng08** [comment]
nah, not that path. try access the **extracted** path

**Comment by JingMatrix** [comment]
Thanks for clarifying your ideas, isolated processes can acess to it.

I will prepare a new pull-request.

**Comment by JingMatrix** [comment]
I try many things to cope with the `AppZygoteInit` problem, but never success.

I am super curious about what makes split APKs method able to solve this problem.

The error logs are given after my commits:
https://github.com/JingMatrix/LSPatch/commit/bc0c92efa6af90d74d33f6d0679180fb72fccae0#commitcomment-115713229

**Comment by JingMatrix** [comment]
After putting the lspatch classes.dex along with the original dex files, I see that the problem:
we cannot hook the  `ActivityThread` started after  `ZygoteInit` to change the classLoader.

Could you tell why we cannot put lspatch classes.dex along with the original dex files?

**Pull request closed without merging**